### DISCUSSION TITLE
I can't load javascript and css

### DISCUSSION THREAD

**Discussion by JMorell** [django-components/django-components#154] [Q&A] [answered]
Hi, I am trying to configure the components.py and I have tried everything in the Media class, but I can't get the css and js to load.

My app path is like this: apps/app_name/components.py

I have tried many combinations but none have worked.

**Comment by EmilStenstrom** [comment]
Hi!

Unfortunately there is not enough information in you question to tell you where things have gone wrong. The recommended location is '[your app]/components/calendar/calendar.css' and '[your app]/components/calendar/calendar.js', and that you point to those locations in your Media class. Does this not work?

**Comment by JMorell** [comment]
I added this on components.py:
class Media:
        css = 'apps/portal/components/subscription-services/style.css'
        js = 'apps/portal/components/subscription-services/script.js'

**Comment by David-Guillot** [comment]
Hi there,

Where are your JS/CSS files located when you run `collectstatic`? This `apps` prefix makes me think Django will have a hard time finding them...

**Comment by JMorell** [comment]
I don't have much idea about static, what command should I execute? I put collectstatic and it warned me that it is going to rewrite the existing files.

I am only using the basic static folder in the root of the server.

**Comment by simkimsia** [comment]
Because I literally tried out the readme yesterday, so my memory is still fresh.

<img width="459" alt="image" src="https://user-images.githubusercontent.com/245021/179163900-7de7fa64-5ba3-448d-b254-9861c6b45fee.png">
This is my Django project layout.

Here's the github repo https://github.com/GreenDeploy-io/example-django-using-hypercomponents

`project`, `hypercomponents` are Django apps. 

I created my `components` folder in the same `app` as the `components.py` under `hypercomponents`.

https://github.com/GreenDeploy-io/example-django-using-hypercomponents/tree/main/hypercomponents

Please make sure your setup is as such.

Next, make sure `[your-app]` is part of INSTALLED_APPS. That was my mistake. It's not very explicitly stated in the README.

https://github.com/GreenDeploy-io/example-django-using-hypercomponents/blob/52328c1847e2b3cbf71372f9eba66588a872272f/config/settings/base.py#L86

Once I added this in, things work for me.

Looking at your path `apps/portal/components/subscription-services/style.css` with the apps prefix, makes me wonder if you will have issues.

**Reply by simkimsia** [reply]
Sorry I realized I was mistaken. I too had an issue with the css and js files to my own surprise. ignore what i wrote here.

**Comment by EmilStenstrom** [comment]
@JMorell I've released a new version of django_components and updated the README with better instructions. Can you get it working now?

**Reply by JMorell** [reply]
Hi, with the updated module, I am getting this error with the loaders set in settings.py

```
Traceback (most recent call last):
  File "/home/proyectos/web/manage.py", line 22, in <module>
    main()
  File "/home/proyectos/web/manage.py", line 18, in main
    execute_from_command_line(sys.argv)
  File "/home/proyectos/venv/lib/python3.9/site-packages/django/core/management/__init__.py", line 446, in execute_from_command_line
    utility.execute()
  File "/home/proyectos/venv/lib/python3.9/site-packages/django/core/management/__init__.py", line 440, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/home/proyectos/venv/lib/python3.9/site-packages/django/core/management/base.py", line 414, in run_from_argv
    self.execute(*args, **cmd_options)
  File "/home/proyectos/venv/lib/python3.9/site-packages/django/core/management/commands/runserver.py", line 74, in execute
    super().execute(*args, **options)
  File "/home/proyectos/venv/lib/python3.9/site-packages/django/core/management/base.py", line 460, in execute
    output = self.handle(*args, **options)
  File "/home/proyectos/venv/lib/python3.9/site-packages/django/core/management/commands/runserver.py", line 111, in handle
    self.run(**options)
  File "/home/proyectos/venv/lib/python3.9/site-packages/django/core/management/commands/runserver.py", line 118, in run
    autoreload.run_with_reloader(self.inner_run, **options)
  File "/home/proyectos/venv/lib/python3.9/site-packages/django/utils/autoreload.py", line 680, in run_with_reloader
    start_django(reloader, main_func, *args, **kwargs)
  File "/home/proyectos/venv/lib/python3.9/site-packages/django/utils/autoreload.py", line 661, in start_django
    reloader.run(django_main_thread)
  File "/home/proyectos/venv/lib/python3.9/site-packages/django/utils/autoreload.py", line 343, in run
    autoreload_started.send(sender=self)
  File "/home/proyectos/venv/lib/python3.9/site-packages/django/dispatch/dispatcher.py", line 176, in send
    return [
  File "/home/proyectos/venv/lib/python3.9/site-packages/django/dispatch/dispatcher.py", line 177, in <listcomp>
    (receiver, receiver(signal=self, sender=sender, **named))
  File "/home/proyectos/venv/lib/python3.9/site-packages/django/template/autoreload.py", line 42, in watch_for_template_changes
    for directory in get_template_directories():
  File "/home/proyectos/venv/lib/python3.9/site-packages/django/template/autoreload.py", line 24, in get_template_directories
    items.update(
  File "/home/proyectos/venv/lib/python3.9/site-packages/django/template/autoreload.py", line 24, in <genexpr>
    items.update(
  File "/home/proyectos/venv/lib/python3.9/site-packages/django/template/loaders/cached.py", line 23, in get_dirs
    yield from loader.get_dirs()
TypeError: 'NoneType' object is not iterable
```

**Reply by EmilStenstrom** [reply]
@JMorell Can you show me how you have set your loaders? The stacktrace does not even mention django_templates? Seems your loader is "None" which should not be possible?

**Reply by JMorell** [reply]
```
TEMPLATES = [
    {
        'BACKEND': 'django.template.backends.django.DjangoTemplates',
        'DIRS': [os.path.join(BASE_DIR, 'templates')],
        # 'APP_DIRS': True,
        'OPTIONS': {
            'context_processors': [
                'django.template.context_processors.debug',
                'django.template.context_processors.request',
                'django.contrib.auth.context_processors.auth',
                'django.contrib.messages.context_processors.messages',
            ],
            'loaders':[(
                'django.template.loaders.cached.Loader', [
                    'django.template.loaders.filesystem.Loader',
                    'django.template.loaders.app_directories.Loader',
                    'django_components.template_loader.Loader',
                ]
            )],
        },
    },
]
```

**Reply by EmilStenstrom** [reply]
Ok. So why is the stacktrace pointing to `/home/proyectos/venv/lib/python3.9/site-packages/django/template/loaders/cached.py`, when that loader isn't even in the list?

**Reply by JMorell** [reply]
this only happens with the changes applied to the file template_loader.py

**Reply by EmilStenstrom** [reply]
@JMorell You mean with the latest version of django_components? What happens when you remove the cached loader from loaders?

**Comment by liquidiert** [comment]
Hey @EmilStenstrom this happens for me too. I still have my custom project structure but using the separated component approach now. Event though I have the load tag for component js dependencies in my template 
`{% component_js_dependencies %}`
 it only outputs a placeholder:
![image](https://github.com/EmilStenstrom/django-components/assets/36642878/bebf56af-aea9-4116-86be-26aa34152f0d)
Can I get you any more information on how to debug this?

**Reply by JuroOravec** [reply]
@liquidiert can you paste your Django settings, especially for STATICFILES_DIRS and TEMPLATES, and the version of django-components you're using?

**Reply by liquidiert** [reply]
@JuroOravec sure!
templates:
```python
TEMPLATES = [
    {
        "BACKEND": "django.template.backends.django.DjangoTemplates",
        "DIRS": [
            # Contains plain text templates, like `robots.txt`:
            BASE_DIR
            / "templates",
        ],
        "OPTIONS": {
            "context_processors": [
                # Default template context processors:
                "django.contrib.auth.context_processors.auth",
                "django.template.context_processors.debug",
                "django.template.context_processors.i18n",
                "django.template.context_processors.media",
                "django.contrib.messages.context_processors.messages",
                "django.template.context_processors.request",
            ],
            "loaders": [
                (
                    "django.template.loaders.cached.Loader",
                    [
                        "django.template.loaders.filesystem.Loader",
                        "django.template.loaders.app_directories.Loader",
                        "django_components.template_loader.Loader",
                    ],
                )
            ],
            "builtins": [
                "django_components.templatetags.component_tags",
            ],
        },
    }
]
```

static files:
```python
STATICFILES_FINDERS = (
    "compressor.finders.CompressorFinder",
    "django.contrib.staticfiles.finders.FileSystemFinder",
    "django.contrib.staticfiles.finders.AppDirectoriesFinder",
)

STATICFILES_DIRS = [
    BASE_DIR / "components",
    BASE_DIR / "static",
]
```

components version: ^0.70

**Reply by JuroOravec** [reply]
Thanks @liquidiert, sorry to ask again, but could you also include setting for django-components (`COMPONENTS`) and your middlewares? And the project folder structure.

Do you have autodiscovery turned off by chance? It should be `True` (default)

Do you have `django_components.middleware.ComponentDependencyMiddleware` among your middlewares? That's the part that currently replaces the `PLACEHOLDER` string with actual content.

If yes to both, and it's still not working, I'm out of my depth. In that case, if you can access the python packages, you can try to drop debug/print statements into `django_components/middleware.py` to debug.

**Reply by liquidiert** [reply]
Ohh @JuroOravec that's embarrassing.... I forgot to add the middleware when I started a new project. My bad!

**Reply by JuroOravec** [reply]
Haha, glad we found the issue!

**Comment by JuroOravec** [comment, marked as answer]
Marking this one as completed as it's quite old already, and the issue shouln't be present anymore. If you have similar issues with newer versions, please open a new discussion / issue.
